federation/pkg/federation-controller/namespace/namespace_controller.go

```diff
@@ -0,0 +1,353 @@
+			err := client.Core().Namespaces().Delete(namespace.Name, &api.DeleteOptions{})
+			return err
+		})
+	return nc
+}
+
+func (nc *NamespaceController) Start() {
+	nc.namespaceInformerController.Run(nc.namespaceStopChan)
+	nc.namespaceFederatedInformer.Start()
+
+	// Handle cluster available/unavailable events.
+	go func() {
+		for {
+			select {
+			// All messages on this chanel mean that all namespaces should be reconciliated.
+			case <-nc.clusterDeliverer.GetTargetChanel():
+				nc.reconcileNamespacesOnClusterChange()
+			case <-nc.clusterReconcileStopChan:
+				return
+			}
+		}
+	}()
+
+	// Handle namespace reconciliation.
+	go func() {
+		for {
+			select {
+			case item := <-nc.namespaceDeliverer.GetTargetChanel():
+				ni := item.Value.(*namespaceItem)
+				nc.reconcileNamespace(ni.namespace, ni.trial)
+			case <-nc.namespaceReconcileStopChan:
+				return
+			}
+		}
+	}()
+}
+
+func (nc *NamespaceController) Stop() {
+	nc.namespaceFederatedInformer.Stop()
+	close(nc.namespaceStopChan)
+	close(nc.clusterReconcileStopChan)
+	close(nc.namespaceReconcileStopChan)
+}
+
+func (nc *NamespaceController) deliverNamespaceObj(obj interface{}, delay time.Duration, trial int64) {
+	namespace, ok := obj.(*api.Namespace)
+	if ok {
+		nc.deliverNamespace(namespace.Name, delay, trial)
+	}
+}
+
+func (nc *NamespaceController) deliverNamespace(namespace string, delay time.Duration, trial int64) {
+	nc.namespaceDeliverer.DeliverAt(
+		namespace,
+		&namespaceItem{
+			namespace: namespace,
+			trial:     trial},
+		time.Now().Add(delay))
+}
+
+func (nc *NamespaceController) isSynced() bool {
+	clusters, err := nc.namespaceFederatedInformer.GetReadyClusters()
+	if err != nil {
+		glog.Errorf("Failed to get ready clusters: %v", err)
+		return false
+	}
+	if !nc.namespaceFederatedInformer.ClustersSynced() {
+		glog.V(2).Infof("Cluster list not synced")
+		return false
+	}
+	if !nc.namespaceFederatedInformer.GetTargetStore().ClustersSynced(clusters) {
+		return false
+	}
+	return true
+}
+
+func (nc *NamespaceController) reconcileNamespacesOnClusterChange() {
+	if !nc.isSynced() {
+		nc.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(ClusterAvailableDelay))
+	}
+	for _, obj := range nc.namespaceInformerStore.List() {
+		namespace := obj.(*api_v1.Namespace)
+		nc.deliverNamespace(namespace.Name, SmallDelay, 0)
+	}
+}
+
+func backoff(trial int64) time.Duration {
+	if trial > 12 {
+		return 12 * 5 * time.Second
+	}
+	return time.Duration(trial) * 5 * time.Second
+}
+
+func (nc *NamespaceController) reconcileNamespace(namespace string, trial int64) {
+	if !nc.isSynced() {
+		nc.deliverNamespace(namespace, ClusterAvailableDelay, trial)
+	}
+
+	baseNamespaceObj, exist, err := nc.namespaceInformerStore.GetByKey(namespace)
+	if err != nil {
+		glog.Errorf("Failed to query main namespace store for %v: %v", namespace, err)
+		nc.deliverNamespace(namespace, backoff(trial+1), trial+1)
+		return
+	}
+	if !exist {
+		// Not federated namespace, ignoring.
```

Comment:
This may also mean, that the federated namespace had been deleted in the meantime. Do we really want to ignore this case?